geometry/sign.hpp

```diff
@@ -4,41 +4,65 @@
 #include <string>
 
 #include "base/not_null.hpp"
+#include "quantities/quantities.hpp"
 #include "serialization/geometry.pb.h"
 
 namespace principia {
 namespace geometry {
 namespace internal_sign {
 
 using base::not_null;
+using quantities::Quantity;
 
 // An element of the multiplicative group ({+1, -1}, *). Useful for instance to
 // represent the determinant of an orthogonal map.
 class Sign final {
  public:
-  template<typename Scalar> explicit Sign(Scalar const& s);
+  explicit Sign(double x);
+  template<typename Dimensions>
+  explicit Sign(Quantity<Dimensions> const& x);
+  // The deleted constructor forbids construction from an integer via
+  // integer-to-double conversion.  Integers have no signed 0, so this could
+  // lead to confusing behaviour.
+  template<typename T>
+  explicit Sign(T x) = delete;
+
+  static constexpr Sign Positive();
+  static constexpr Sign Negative();
+
+  template<typename T, typename = std::enable_if_t<std::is_arithmetic_v<T>>>
+  static constexpr Sign OfNonZero(T x);
+
+  constexpr bool is_positive() const;
+  constexpr bool is_negative() const;
 
-  bool Negative() const;
-  bool Positive() const;
+  constexpr Sign operator+() const;
+  constexpr Sign operator-() const;
 
-  bool operator==(Sign const& other) const;
-  bool operator!=(Sign const& other) const;
+  // Returns ±1.
+  constexpr operator int() const;
+
+  constexpr bool operator==(Sign const& other) const;
+  constexpr bool operator!=(Sign const& other) const;
 
   void WriteToMessage(not_null<serialization::Sign*> message) const;
   static Sign ReadFromMessage(serialization::Sign const& message);
 
  private:
+  constexpr explicit Sign(bool negative);
+
   bool negative_;
-  friend Sign operator*(Sign const& left, Sign const& right);
+
+  friend constexpr Sign operator*(Sign const& left, Sign const& right);
   template<typename T>
-  friend T operator*(Sign const& left, T const& right);
+  friend constexpr T operator*(Sign const& left, T const& right);
 };
 
-Sign operator*(Sign const& left, Sign const& right);
+constexpr Sign operator*(Sign const& left, Sign const& right);
```

Comment:
Why do we have `*` here but `==` and `!=` in the class?  We should probably decide once and for all what we do with binary operators.

Reply:
A job for another pull request, we should harmonize `Quantity` with `Multivector` as well.